src/levanter/models/hyena.py

```diff
@@ -0,0 +1,542 @@
+            y = self.dropout(y, key=dropout_key)
+
+        return y
+
+
+class HyenaOperator(eqx.Module):
+    """Hyena operator - the core building block of the Hyena architecture."""
+
+    config: HyenaConfig = eqx.field(static=True)
+    in_proj: hnn.Linear
+    out_proj: hnn.Linear
+    short_filter: hnn.Conv
+    filter_fn: HyenaFilter
+    dropout: hnn.Dropout
+    activation: ActivationFunction = eqx.field(static=True)
+
+    @staticmethod
+    def init(config: HyenaConfig, *, key):
+        keys = jrandom.split(key, 5)
+
+        in_proj = hnn.Linear.init(
+            In=config.Embed,
+            Out=config.EmbedOrderPlus1,
+            key=keys[0],
+            use_bias=config.use_bias,
+        )
+
+        # Output projection: hidden_size -> hidden_size
+        # Create a new axis with the same dimensions to avoid naming collision
+        # We do not support inner_factor from the PyTorch impl.
+        out_proj = hnn.Linear.init(
+            In=config.Embed, Out=config.Embed.alias("output_embed"), key=keys[1], use_bias=config.use_bias
+        )
+
+        short_filter = hnn.Conv.init(
+            Spatial=config.Pos,
+            In=config.EmbedOrderPlus1,
+            Out=config.EmbedOrderPlus1.alias("out_channels"),
+            kernel_size=config.short_filter_order,
+            groups=config.EmbedOrderPlus1.size,
+            padding=config.short_filter_order - 1,
+            key=keys[2],
+        )
+
+        filter_fn = HyenaFilter.init(config, key=keys[3])
+        dropout = hnn.Dropout(pdrop=config.resid_pdrop)
+
+        return HyenaOperator(
+            config=config,
+            in_proj=in_proj,
+            out_proj=out_proj,
+            short_filter=short_filter,
+            filter_fn=filter_fn,
+            dropout=dropout,
+            activation=config.activation.to_fn(),
+        )
+
+    @named_call
+    def __call__(self, u: hax.NamedArray, *, key: PRNGKeyArray | None = None) -> hax.NamedArray:
+        key_in_proj, key_dropout = haliax.jax_utils.maybe_rng_split(key, 2)
+        Pos = self.config.Pos
+        Block = self.config.Block
+        PosPerBlock = self.config.PosPerBlock
+        EmbedOrderPlus1 = self.config.EmbedOrderPlus1
+        Embed = self.config.Embed
+        # input has the same axis name as the Pos axis, but possibly different size
+        input_length = u.axis_size(Pos.name)
+        l_filter = min(input_length, Pos.size)
+
+        # Input projection from [Embed] to [(order+1) * Embed]
+        u = self.in_proj(u, key=key_in_proj)
+
+        # trying to keep the variable names from the official impl.
+        # I think uc stands for "u convolved".
+        uc = self.short_filter(u).rename({"out_channels": EmbedOrderPlus1})
+        # Note: after the short filter, uc."position" axis has slightly larger size due to padding.
+        # Hence we use Pos.name rather than Pos to slice, otherwise haliax complains the axis sizes
```

Comment:
what do you think about this behavior? I think I don't like it anymore and am thinking of changing it